lambeq/backend/quantum.py

```diff
@@ -1162,3 +1164,482 @@ def CRz(phi, distance=1): return Controlled(Rz(phi), distance)  # noqa: E731
+                                     gate_layer.right))
+                    new_gates.append(Layer(gate_layer.left,
+                                     Swap(qubit, qubit),
+                                     dom >> gate_layer.right))
+
+        return new_gates, q_idx
+
+    def build_left_right(q_idx, layer, layers):
+        """
+        We assume that the left and right are constructable
+        from the last gate
+        and the left position of the bra.
+        (We type check at the end.)
+        Rebuild left and right based on the last layer
+        """
+        if len(layers) == 0:
+            return layer
+
+        gate_layer = layers[-1]
+
+        total_layer = ([*gate_layer.left] + [*gate_layer.box.cod]
+                       + [*gate_layer.right])
+
+        # Assumes you're only inserting one qubit at a time
+        total_layer[q_idx] = layer.box.cod
+
+        if q_idx == 0 or not total_layer[:q_idx]:
+            layer.left = Ty()
+        else:
+            layer.left = layer.left._fromiter(total_layer[:q_idx])
+
+        if q_idx == len(total_layer) - 1 or not total_layer[q_idx+1:]:
+            layer.right = Ty()
+        else:
+            layer.right = layer.right._fromiter(total_layer[q_idx+1:])
+
+        return layer
+
+    circuit = Functor(target_category=quantum,  # type: ignore [assignment]
+                      ob=lambda _, x: x,
+                      ar=remove_ketbra1)(circuit)  # type: ignore [arg-type]
+
+    layers = circuit.layers
+
+    for i, layer in enumerate(layers):
+
+        if isinstance(layer.box, Ket):
+            qubits, gates = add_qubit(qubits,
+                                      layer,
+                                      len(layer.left),
+                                      gates)
+
+        elif isinstance(layer.box, (Bra, Discard)):
+
+            q_idx = len(layer.left)
+            layers[i+1:], q_idx = pull_qubit_through(q_idx, layers[i+1:])
+            layer = build_left_right(q_idx, layer, layers[i+1 :])
+
+            postselect.insert(0, layer)
+
+        else:
+            gates.append(layer)
+
+    if gates:
+        postselect = construct_measurements(gates[-1], postselect)
+
+    # Rebuild the diagram
+    diags = [Diagram(dom=layer.dom, cod=layer.cod, layers=[layer])  # type: ignore [arg-type] # noqa: E501
+             for layer in qubits + gates + postselect + measures]
+
+    layerD = diags[0]
+    for diagram in diags[1:]:
+        layerD = layerD >> diagram
+
+    return layerD
+
+
+def circuital_to_dict(diagram: Diagram):
+    """Takes a circuital :py:class:`lambeq.quantum.Diagram`, returns
+    a dictionary for converting to a circuit.
+    Will check if the diagram is circuital before converting.
+
+    Parameters
+    ----------
+    diagram : :py:class:`~lambeq.backend.quantum.Diagram`
+        The :py:class:`Circuits <lambeq.backend.quantum.Diagram>`
+        to be converted to dictionary.
+
+    Returns
+    -------
+    Dict:
```

Comment:
Let's create a dedicated re-usable public class for the circuit dictionary, with specific attributes for each part/list. We might add more functionality in the future.